### PR TITLE
docs(block): add docs about style inheritance

### PR DESCRIPTION
**Pull request by joshka**
Fixes: https://github.com/ratatui-org/ratatui/issues/1129

### CODE REVIEW

File: src/widgets/block.rs
Comment (EdJoPaTo) [inline src/widgets/block.rs:426]: I think the sentence in from of this should be phrased more clearly, but I don't have a good idea how yet.
Currently, it reads like `border_style` overrides, but actually the `.not_italic()` overrides.

File: src/widgets/block.rs
Comment (EdJoPaTo) [inline src/widgets/block.rs:41]: This is the case here too. `overridden` how? By another style on top? no, by another style with reset or actively setting what to remove (`no_…`).

File: src/widgets/block.rs
Comment (joshka) [reply src/widgets/block.rs:426]: Funnily, the user that was originally confused about this implementation stated they understood the explanation well. 😀.

The styles are actually overridden as specified, style by style, not element by element.

Comment (joshka) [reply src/widgets/block.rs:426]: Ah I can see how what I wrote is possibly confusing. The way that one style overrides another is by combination of the parts that are specified overwriting those that are previously specified. That's the "overriding" that I meant here.

This is called our more explicitly in [`Style`](https://docs.rs/ratatui/0.26.3/ratatui/style/struct.Style.html):

> Styles represents an incremental change. If you apply the styles S1, S2, S3 to a cell of the terminal buffer, the style of this cell will be the result of the merge of S1, S2 and S3, not just S3.

Comment (joshka) [reply src/widgets/block.rs:428]: Added comment in 1e7b6f47a

File: src/widgets/block.rs
Comment (EdJoPaTo) [inline src/widgets/block.rs:423]: The example is way more helpful than the current sentence here. I still think override is a confusing word. unset styles or add partially to override seems better to me, as that doesnt imply the style "below" is ignored then.
Alternatively, remove the sentence in front of this code block completely as the description above explains them to be applied on top of the block style already.

```suggestion
    /// // For border and title you can partially override / unset styles set on the block level.
```

**Comment by joshka** [comment]
ping @ratatui-org/maintainers

File: src/widgets/block.rs
Comment (kdheepak) [inline src/widgets/block.rs:413]: At some point I think we should rephrase this? This says "inside that block" and "inner widget" but that isn't how the code works and I don't think we use that language on the website docs.

**Review by kdheepak — APPROVED** [review]

File: src/widgets/block.rs
Comment (joshka) [reply src/widgets/block.rs:413]: Can you suggest alternate wording for this?

Comment (kdheepak) [reply src/widgets/block.rs:413]: The shortest change I can think of is:

```suggestion
    /// This will also apply to the widget that houses the block, in which case the widget's styles will be applied on top of the block's styles.
```

But I'm fine with scoping this PR to not include this kind of change because I think we'd have to read through all the docs with an eye to make them consistent.

Comment (kdheepak) [reply src/widgets/block.rs:413]: Hmm, isn't the original docs and the change in this PR (and my previous comment suggestion) all incorrect?

For a `Paragraph`:

https://github.com/ratatui-org/ratatui/blob/cbfcd34ef5cfaa28f72706af373b7dfe4a648795/src/widgets/paragraph.rs#L332-L334

The code for a `Paragraph` applies the `block` style _after_ applying the paragraph's own style. I always thought it was the other way around. I haven't checked all the widgets for consistency here.

Comment (kdheepak) [reply src/widgets/block.rs:413]: The `block`'s style overlaps or is "painted over" the paragraph's style. In the actual behavior, the block's style "overrides" the paragraph's style. 

I also think "overrride" implies that the block's style replaces the paragraph's style but really the block's style is superimposed on top of the paragraph's style. A paragraph style with background white and a block style with foreground red will result in "red on white" behavior, right?

Comment (joshka) [reply src/widgets/block.rs:413]: This comment refers to widgets that are rendered in the Block::inner rect, not in widgets that have blocks.

> The code for a Paragraph applies the block style after applying the paragraph's own style. I always thought it was the other way around. I haven't checked all the widgets for consistency here.

This is the consistent behavior for all widgets that have a block and widget style. If this was the other way around, then styling borders would be impossible because the styles set on the widget would override the block. I have a vague recollection this is covered in an issue / PR comment from eons ago, but the behavior is correct either way.

> I also think "overrride" implies that the block's style replaces the paragraph's style but really the block's style is superimposed on top of the paragraph's style. A paragraph style with background white and a block style with foreground red will result in "red on white" behavior, right?

Override here means the same as what we're taking about in terms of merging in the style docs:

> Styles represents an incremental change. If you apply the styles S1, S2, S3 to a cell of the terminal buffer, the style of this cell will be the result of the merge of S1, S2 and S3, not just S3.

The problem with the word merge is that it doesn't imply a precedence of the items that are merged, whereas override does imply this.

Comment (kdheepak) [reply src/widgets/block.rs:413]: > This comment refers to widgets that are rendered in the Block::inner rect, not in widgets that have blocks.

My bad, I don't think I caught that it is referring to `Block::inner`. I still don't quite follow what this line is supposed to mean then. The following line was just above it, so I read it as widgets that have a block.

```rust
/// If the block is used as a container for another widget, the inner widget can also be styled.
```

> The problem with the word merge is that it doesn't imply a precedence of the items that are merged, whereas override does imply this.

Maybe "overriding merge" would solve my nit here.

Comment (kdheepak) [reply src/widgets/block.rs:413]: None of my comments are blocking this PR!

Comment (joshka) [reply src/widgets/block.rs]: tweaked it a little 05389d5b8

**Pull request merged**
